File: src/main/java/dev/shota/decompiler/window/Window.java
package dev.shota.decompiler.window;

import com.formdev.flatlaf.util.SystemInfo;
import com.sun.javafx.tk.Toolkit;
import dev.shota.decompiler.loader.FileLoader;
import dev.shota.decompiler.reflection.Instance;
import dev.shota.decompiler.reflection.Singleton;
import dev.shota.decompiler.window.explorer.Explorer;
import dev.shota.decompiler.window.menu.MenuBar;
import dev.shota.decompiler.window.utils.Styles;
import dev.shota.decompiler.window.viewer.Viewer;
import javafx.embed.swing.JFXPanel;
import javafx.scene.Scene;
import javafx.scene.control.SplitPane;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.Pane;
import javafx.scene.layout.Region;
import javafx.scene.text.Font;
import lombok.SneakyThrows;
import org.jetbrains.annotations.NotNull;
import org.reflections.Reflections;
import org.reflections.scanners.Scanners;
import javax.swing.*;
import java.awt.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.dnd.*;
import java.io.File;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Singleton
public class Window extends JFrame implements DropTargetListener {

    @SneakyThrows
    private Window() {
        Set<String> fonts = new Reflections("fonts", Scanners.Resources).getResources(Pattern.compile(".*\\.ttf"));
        for (String font : fonts)
            Font.loadFont(getClass().getClassLoader().getResourceAsStream(font), Toolkit.getToolkit().getFontLoader().getSystemFontSize());

        Region explorer = Instance.get(Explorer.class);
        Region viewer = Instance.get(Viewer.class);
        BorderPane sidebar = new BorderPane(explorer);
        SplitPane root = new SplitPane(sidebar, viewer);
        SplitPane.setResizableWithParent(sidebar, false);
        root.getStylesheets().addAll(Styles.get("main"), Styles.get("code"));
        root.setDividerPositions(
                explorer.getMinWidth() / (explorer.getMinWidth() + viewer.getMinWidth()),
                viewer.getMinWidth() / (explorer.getMinWidth() + viewer.getMinWidth())
        );

        if (SystemInfo.isMacOS && SystemInfo.isMacFullWindowContentSupported) {
            rootPane.putClientProperty("apple.awt.fullWindowContent", true);
            rootPane.putClientProperty("apple.awt.transparentTitleBar", true);
            rootPane.putClientProperty("apple.awt.windowTitleVisible", false);
            rootPane.putClientProperty("apple.awt.fullscreenable", true);
            Pane space = new Pane();
            space.setPrefHeight(25);
            sidebar.setTop(space);
        }

        Image logo = new ImageIcon(Objects.requireNonNull(getClass().getClassLoader().getResourceAsStream("logo/logo.png")).readAllBytes()).getImage();
        if (Taskbar.isTaskbarSupported() && Taskbar.getTaskbar().isSupported(Taskbar.Feature.ICON_IMAGE))
            Taskbar.getTaskbar().setIconImage(logo);
        setIconImage(logo);

        DropTarget dropTarget = new DropTarget();
        dropTarget.addDropTargetListener(this);
        setDropTarget(dropTarget);

        JFXPanel panel = new JFXPanel();
        panel.setScene(new Scene(root, 894, 528));
        panel.getDropTarget().setActive(false);
        add(panel);
        pack();
        setTitle("Decompiler");
        setDefaultCloseOperation(DISPOSE_ON_CLOSE);
        setJMenuBar(new MenuBar());
        setLocationRelativeTo(null);
    }

    @Override
    public void dispose() {
        super.dispose();
        System.exit(0);
    }

    @Override
    public void dragEnter(@NotNull DropTargetDragEvent event) {
        event.acceptDrag(DnDConstants.ACTION_COPY);
    }

    @Override
    public void dragOver(@NotNull DropTargetDragEvent event) {
        event.acceptDrag(DnDConstants.ACTION_COPY);
    }

    @Override
    @SneakyThrows
    public void drop(@NotNull DropTargetDropEvent event) {
        event.acceptDrop(DnDConstants.ACTION_COPY);
        if (event.getTransferable().isDataFlavorSupported(DataFlavor.javaFileListFlavor)) {
            List<?> list = (List<?>) event.getTransferable().getTransferData(DataFlavor.javaFileListFlavor);
            List<File> files = list.stream().map(o -> (File) o).collect(Collectors.toList());
            event.dropComplete(FileLoader.load(files));
            return;
        }
        event.dropComplete(false);
    }

    @Override
    public void dropActionChanged(DropTargetDragEvent event) {}
    @Override
    public void dragExit(DropTargetEvent event) {}

}

File: src/main/kotlin/dev/shota/decompiler/window/Window.kt
package dev.shota.decompiler.window

import com.formdev.flatlaf.util.SystemInfo
import dev.shota.decompiler.loader.FileLoader
import dev.shota.decompiler.window.explorer.Explorer
import dev.shota.decompiler.window.menu.MenuBar
import dev.shota.decompiler.window.utils.Styles
import dev.shota.decompiler.window.viewer.Viewer
import javafx.embed.swing.JFXPanel
import javafx.scene.Scene
import javafx.scene.control.SplitPane
import javafx.scene.layout.BorderPane
import javafx.scene.layout.Pane
import java.awt.Dimension
import java.awt.Taskbar
import java.awt.datatransfer.DataFlavor
import java.awt.dnd.*
import java.io.File
import java.util.*
import java.util.stream.Collectors
import javax.swing.ImageIcon
import javax.swing.JFrame
import kotlin.system.exitProcess

object Window : JFrame(), DropTargetListener {

    init {
        val sidebar = BorderPane(Explorer)
        val root = SplitPane(sidebar, Viewer)
        SplitPane.setResizableWithParent(sidebar, false)
        root.stylesheets.add(Styles.get("main"))
        root.setDividerPositions(Explorer.minWidth / (Explorer.minWidth + Viewer.minWidth), Viewer.minWidth / (Explorer.minWidth + Viewer.minWidth))

        if (SystemInfo.isMacOS && SystemInfo.isMacFullWindowContentSupported) {
            rootPane.putClientProperty("apple.awt.fullWindowContent", true)
            rootPane.putClientProperty("apple.awt.transparentTitleBar", true)
            rootPane.putClientProperty("apple.awt.windowTitleVisible", false)
            rootPane.putClientProperty("apple.awt.fullscreenable", true)
            Pane().run {
                prefHeight = 25.0
                sidebar.top = this
            }
        }

        val logo = ImageIcon(Objects.requireNonNull(javaClass.classLoader.getResourceAsStream("logo/logo.png")).readAllBytes()).image
        if (Taskbar.isTaskbarSupported() && Taskbar.getTaskbar().isSupported(Taskbar.Feature.ICON_IMAGE)) Taskbar.getTaskbar().iconImage = logo
        iconImage = logo

        dropTarget = DropTarget().also {
            it.addDropTargetListener(this)
        }

        JFXPanel().let {
            it.scene = Scene(root, 894.0, 528.0)
            it.dropTarget.isActive = false
            add(it)
        }

        title = "Decompiler"
        defaultCloseOperation = DISPOSE_ON_CLOSE
        jMenuBar = MenuBar()
        minimumSize = Dimension((Explorer.minWidth + Viewer.minWidth).toInt(), ((Explorer.minWidth + Viewer.minWidth) / 2).toInt())
        pack()
        setLocationRelativeTo(null)
    }

    override fun dispose() {
        super.dispose()
        exitProcess(0)
    }

    override fun dragEnter(event: DropTargetDragEvent?) {
        event?.acceptDrag(DnDConstants.ACTION_COPY)
    }

    override fun dragOver(event: DropTargetDragEvent?) {
        event?.acceptDrag(DnDConstants.ACTION_COPY)
    }

    override fun drop(event: DropTargetDropEvent?) {
        event!!.acceptDrop(DnDConstants.ACTION_COPY)
        if (event.transferable.isDataFlavorSupported(DataFlavor.javaFileListFlavor)) {
            val list = event.transferable.getTransferData(DataFlavor.javaFileListFlavor) as List<*>
            val files = list.stream().map { it as File }.collect(Collectors.toList())
            event.dropComplete(FileLoader.load(files))
            return
        }
        event.dropComplete(false)
    }

    override fun dropActionChanged(event: DropTargetDragEvent?) {}
    override fun dragExit(event: DropTargetEvent?) {}

}